### ISSUE TITLE
Add option for custom

### ISSUE THREAD

**Issue by ionrover2** [jc21/nginx-proxy-manager#1417]
**Is your feature request related to a problem? Please describe.**
I would like this tool to be used with a self signed CA in an internal environment that has an acme compliant server. I'm currently using Step CA for my acme compliant server. I currently don't have the option to use my own acme server without a ton of involved reconfiguration.

**Describe the solution you'd like**
An advanced configuration option where you can upload a self signed CA and give a custom acme compliant url to be used with a given host.

**Describe alternatives you've considered**
I am currently using a combination of jwilder/nginx-proxy and the encrypt companion to accomplish the same task with a custom acme CA, but it would be ideal to have a graphical frontend for a reverse proxy that can pull a valid internal cert, route traffic to separate physical hosts as needed and also the underlying containers running on the same machine.

**Additional context**

**Comment by NetHero-es** [comment]
Are there any plans to integrate this feature? I'm having the exact same requirement.

**Comment by francescocaponio** [comment]
I'm using NPM both for intranet traffic and for traffic coming from outside. For the latter, it works amazingly as it is, for the internal, would be nice if it could work with Smallstep CA (ACME and compatible with certbot) and avoid the manual certificate renewal every year for each service!

**Comment by SantaSpeen** [comment]
+1

**Comment by lriley2020** [comment]
Would love it if this feature was added!

**Comment by vshaev** [comment]
+1

**Comment by charliemaiors** [comment]
+1

**Comment by francescocaponio** [comment]
So, this is my temporary solution:

I create my own container with this Dockerfile for each new version of NPM:

```
FROM docker.io/jc21/nginx-proxy-manager:2.11.1

RUN curl http://<internal_http_server>.lan/rootca.crt -o rootca.crt \
    && mv rootca.crt /usr/local/share/ca-certificates/rootca.crt \
    && update-ca-certificates

RUN echo -e "\nserver=https://labca.lan/directory\n" >> /etc/letsencrypt.ini

RUN sed -i "s\Let's Encrypt\LabCa\g" /app/frontend/js/*.js
```

This custom container can only work with the internal CA instance and not with Let's Encrypt. You can't have both. If you need also external domain you must run a second instance of NPM.

After this, I run the container in the same way I did before with the original NPM container, but I'm able to create and renew internal certificates with [LabCA](https://github.com/hakwerk/labca/):

<img width="949" alt="image" src="https://github.com/NginxProxyManager/nginx-proxy-manager/assets/66728654/73ab96d5-70bc-4ce6-bd0e-6e24c883807b">

**Comment by accessiblepixel** [comment]
> FROM docker.io/jc21/nginx-proxy-manager:2.11.1
> 
> RUN curl http://<internal_http_server>.lan/rootca.crt -o rootca.crt \
>     && mv rootca.crt /usr/local/share/ca-certificates/rootca.crt \
>     && update-ca-certificates
> 
> RUN echo -e "\nserver=https://labca.lan/directory\n" >> /etc/letsencrypt.ini
> 
> RUN sed -i "s\Let's Encrypt\LabCa\g" /app/frontend/js/*.js

Hmm. I tried using your solution @francescocaponio but whenever I try to issue a new certificate I get:
``Certificate is not valid (Cannot read properties of null (reading '1'))``

I'm using stepca as the bases for my ACME server. According to the logs from stepca the certificate gets issued but it seems that nginxproxymanager UI doesn't know what to do with it. Don't suppose you have any ideas?

From the logs it looks like it does indeed get issued...
```
app_1  | Successfully received certificate.
app_1  | Certificate is saved at: /etc/letsencrypt/live/npm-9/fullchain.pem
app_1  | Key is saved at:         /etc/letsencrypt/live/npm-9/privkey.pem
app_1  | This certificate expires on 2024-03-09.
app_1  | These files will be updated when the certificate renews.
```

**Comment by francescocaponio** [comment]
Hi,
I also have the same log 
```
Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/npm-15/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/npm-15/privkey.pem
This certificate expires on 2024-06-04.
These files will be updated when the certificate renews.
NEXT STEPS:
...
```
try to analyze the content of the file `/tmp/letsencrypt-log/letsencrypt.log` of the npm container for a DEBUG level log of certbot, and check also the stepca logs to understand better where the problem could be.

**Comment by oldboys92** [comment]
@accessiblepixel

> Hmm. I tried using your solution @francescocaponio but whenever I try to issue a new certificate I get: `Certificate is not valid (Cannot read properties of null (reading '1'))`
> 
> I'm using stepca as the bases for my ACME server. According to the logs from stepca the certificate gets issued but it seems that nginxproxymanager UI doesn't know what to do with it. Don't suppose you have any ideas?

The problem is caused by npm `certbot`, which after spending some hours I've just discovered that it removes the `subject` in the CSR. Also the default Step CA ACME provisioner configuration is not adding a `CN`/`subject`, when the CSR has none and this is what the problem is.

To solve you need to add

```JSON
  "forceCN": true,
```

in your step-ca config `ca.json`.

Mine looks now like this and certificates are issued without issues.

```JSON
                        {
                                "type": "ACME",
                                "name": "acme",
                                "forceCN": true
                        }
```

Hope this helps you.

**Comment by oldboys92** [comment]
@francescocaponio thank you for your cool solution, the workaround is working like a charm 🖖

PS: it cost me some time, till I realized your `sed` had a small typo. You used `\` instead of `/` for the regex.

**Comment by lriley2020** [comment]
No disrespect to the project, but if you're facing these kinds of challenges, you might be starting to outgrow nginx-proxy-manager! It might be much easier in this scenario to just switch to plain old nginx, or another reverse proxy (eg: caddy). I just switched to caddy a few days ago for the better custom ACME provider support (and several other more advanced features). What you're trying to achieve here can be done in a few lines in a caddyfile:
```
my.domain.com {
    tls {
        issuer acme {
            dir https://ca.domain.com:8443/acme/acme/directory
	    trusted_roots /data/caddy/custom-certs/step-ca/step-ca.pem
        }
    }
    reverse_proxy * http://myappserver.local:8080
}


```
That's it! The whole configuration!

**Comment by oldboys92** [comment]
> No disrespect to the project, but if you're facing these kinds of challenges, you might be starting to outgrow nginx-proxy-manager! It might be much easier in this scenario to just switch to plain old nginx, or another reverse proxy (eg: caddy). I just switched to caddy a few days ago for the better custom ACME provider support (and several other more advanced features). What you're trying to achieve here can be done in a few lines in a caddyfile:

@lriley2020 you are absolutely right, I had same thoughts too. I will probably switch myself to Caddy, but at least for NPM I wanted to see if there is a chance to get ACME working with another RA/CA, since it looked that, not much was missing. And in my case, the package was already there and was convenient to just use it.

**Comment by vshaev** [comment]
+1

**Comment by gpend** [comment]
+1

**Comment by dwydler** [comment]
+1

**Comment by mrkhachaturov** [comment]
+1

**Comment by be-reich** [comment]
+1

**Comment by SubniC** [comment]
+1

**Comment by tofkamp** [comment]
+1^9999999999

**Comment by GLeurquin** [comment]
Nginx introduced native ACME support: https://blog.nginx.org/blog/native-support-for-acme-protocol

This will probably be useful in implementing this feature.

**Comment by t984447** [comment]
+1
Looking around for alternatives JUST for this feature but would love it if NPM would allow for ACME towards custom CA .

**Comment by AlexInABox** [comment]
+1
[francesco's](https://github.com/NginxProxyManager/nginx-proxy-manager/issues/1417#issuecomment-1979284055) custom Dockerfile solution worked perfectly for me using step-ca as my local CA, but I would definitely be more comfortable with an out of the box solution for the nginx-proxy-manager directly!
